File: 2020/Day08/day08.cpp
#include <utility>
#include <iostream>
#include <string>
#include <vector>

struct Line {
    std::string opr;
    int arg;
    bool visited = false;

    Line(std::string opr, int arg) : opr(std::move(opr)), arg(arg) {}
};

std::vector<Line> lines;

int rec(int i = 0, int acc = 0, bool use = false) {
    if (i >= lines.size())
        return acc;
    if (lines[i].visited)
        return 0;
    int res, tmp;
    lines[i].visited = true;
    if (lines[i].opr == "acc")
        res = rec(i + 1, acc + lines[i].arg, use);
    else if (lines[i].opr == "nop") {
        tmp = rec(i + 1, acc, use);
        res = tmp ? tmp : (not use ? rec(i + lines[i].arg, acc, true) : 0);
    } else {
        tmp = rec(i + lines[i].arg, acc, use);
        res = tmp ? tmp : (not use ? rec(i + 1, acc, true) : 0);
    }
    lines[i].visited = false;
    return res;
}

int main() {
    std::string opr;
    int arg, acc{}, i{};
    while (std::cin >> opr >> arg)
        lines.emplace_back(opr, arg);
    while (not lines[i].visited) {
        lines[i].visited = true;
        if (lines[i].opr == "acc") {
            acc += lines[i].arg;
            i++;
        } else if (lines[i].opr == "nop")
            i++;
        else
            i += lines[i].arg;
    }
    for (auto &line: lines)
        line.visited = false;
    std::cout << acc << '\n' << rec() << '\n';
}

File: 2020/Day08/day08.py
def rec(i=0, acc=0, use=False):
    if i >= n:
        return acc
    if vlines[i][1]:
        return None
    vlines[i][1] = True
    opr, arg = vlines[i][0].split()
    if opr == 'acc':
        res = rec(i + 1, acc + int(arg), use)
    elif opr == 'nop':
        res = rec(i + 1, acc, use) or (rec(i + int(arg), acc, True) if not use else None)
    else:
        res = rec(i + int(arg), acc, use) or (rec(i + 1, acc, True) if not use else None)
    vlines[i][1] = False
    return res


lines = open('inp08').readlines()
n = len(lines)
vlines = [[line, False] for line in lines]
acc = i = 0
while lines[i]:
    opr, arg = lines[i].split()
    lines[i] = ''
    if opr == 'nop':
        i += 1
    elif opr == 'acc':
        acc += int(arg)
        i += 1
    else:
        i += int(arg)

print(acc, rec(), sep='\n')
